frontend/lib/resolution/scope-queries.cpp

```diff
@@ -2244,6 +2262,306 @@ void resolveUsesAndImportsInScope(Context* context, const Scope* scope) {
+                               int& nValue,
+                               int& nConstRef,
+                               int& nRef,
+                               int& nOther) {
+  nValue = 0;
+  nConstRef = 0;
+  nRef = 0;
+  nOther = 0;
+  for (const auto& b : v) {
+    for (const auto& id : b) {
+      Function::ReturnIntent reti = parsing::idToFnReturnIntent(context, id);
+      switch (reti) {
+        case Function::DEFAULT_RETURN_INTENT:
+        case Function::CONST:
+          nValue++;
+          break;
+        case Function::CONST_REF:
+          nConstRef++;
+          break;
+        case Function::REF:
+          nRef++;
+          break;
+        case Function::PARAM:
+        case Function::TYPE:
+          nOther++;
+          break;
+      }
+    }
+  }
+}
+
+
+static const bool&
+emitMultipleDefinedSymbolErrorsQuery(Context* context, const Scope* scope) {
+  QUERY_BEGIN(emitMultipleDefinedSymbolErrorsQuery, context, scope);
+
+  bool result = false;
+
+  std::set<UniqueString> namesDefined;
+  std::set<UniqueString> namesDefinedMultiply;
+  ScopeSet checkedScopes;
+
+  collectAllNames(context, scope, /* skip private */ false,
+                  namesDefined, namesDefinedMultiply, checkedScopes);
+
+  // Now, consider names in namesDefinedMultiply. If there are any
+  // that are not only parenful functions, issue an error.
+  LookupConfig config = LOOKUP_DECLS |
+                        LOOKUP_IMPORT_AND_USE |
+                        LOOKUP_SKIP_SHADOW_SCOPES;
+  for (auto name : namesDefinedMultiply) {
+    std::vector<BorrowedIdsWithName> v =
+      lookupNameInScope(context, scope, { }, name, config);
+    int nParenfulMethods = 0;
+    int nParenfulNonMethodFunctions = 0;
+    int nParenlessMethods = 0;
+    int nParenlessNonMethodFunctions = 0;
+    int nFields = 0;
+    int nOther = 0;
+    countSymbols(context, v,
+                 nParenfulMethods, nParenfulNonMethodFunctions,
+                 nParenlessMethods, nParenlessNonMethodFunctions,
+                 nFields, nOther);
+    int nNonFunctions = nFields + nOther;
+    bool error = false;
+    if (nParenfulNonMethodFunctions > 0 &&
+        nParenlessNonMethodFunctions + nNonFunctions > 0) {
+      // mix of parenful functions and parenless functions / non-function
+      error = true;
+    } else if (nParenlessNonMethodFunctions > 0 && nNonFunctions  > 0) {
+      // mix of parenless functions and non-functions
+      error = true;
+    } else if (nNonFunctions > 1) {
+      // multiple non-functions with the same name
+      error = true;
+    } else if (nParenlessNonMethodFunctions > 1) {
+      // multiple parenless non-method functions with the same name
+      // check: is it return intent overloading?
+      int nValue = 0;
+      int nConstRef = 0;
+      int nRef = 0;
+      int nOther = 0;
+      countReturnIntents(context, v, nValue, nConstRef, nRef, nOther);
+      if (nValue <= 1 && nConstRef <= 1 && nRef <= 1 && nOther == 0) {
```

Comment:
Why can there be no _other_ (`param` or `type`) return intent parenless functions in a valid overload?
(I wasn't able to answer this question with the [spec](https://chapel-lang.org/docs/language/spec/procedures.html#the-param-return-intent))

Reply:
It is covered here https://chapel-lang.org/docs/language/spec/procedures.html#choosing-return-intent-overloads-based-on-calling-context -- "The compiler can choose between overloads differing in return intent when:" covers value, `const ref`, and `ref` return intent; but not `param` or `type` return intents. So for those, we can read it to mean that the compiler can't choose between the overloads (meaning that there will be an ambiguity error).

Reply:
(But the real answer is that we haven't had a motivating case to change the language to allow return intent overloading based on `param` or `type` return intent. AFAIK such a change can be non-breaking and can be done after 2.0, though).